File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

bool can_convert(string word1, string word2) {
    vector<bool> check(word2.length(), false);
    int cnt = 0;
    bool is_duplicate;
    for (char c: word1) {
        is_duplicate = false;
        for (int i = 0; i < word2.length(); i++) {
            if (c == word2[i] && !check[i]) {
                check[i] = true;
                is_duplicate = true;
                break;
            }
        }
        if (!is_duplicate) cnt++;
        if (cnt > 1) return false;
    }
    return cnt == 1;
}

vector<vector<bool>> get_graph(vector<string>& words) {
    int n = words.size();
    vector<vector<bool>> graph(n, vector<bool>(n, false));
    for (int i = 0; i < n - 1; i++) {
        for (int j = i + 1; j < n; j++) {
            if (can_convert(words[i], words[j])) {
                graph[i][j] = true;
                graph[j][i] = true;
            }
        }
    }
    return graph;
}

int bfs(vector<vector<bool>>& graph, vector<int>& visited, 
        vector<string>& words, string target) {
    // begin 단어 노드 방문 처리
    queue<int> q; q.push(words.size() - 1);
    visited[words.size() - 1] = 0;
    
    while (!q.empty()) {
        int t = q.front(); q.pop();
        if (words[t] == target) {
            return visited[t];
        }
        for (int i = 0; i < graph.size(); i++) {
            if (graph[t][i] && visited[i] == -1) {
                visited[i] = visited[t] + 1;
                q.push(i);
            }
        }
    }
    return 0;
}

int solution(string begin, string target, vector<string> words) {
    words.push_back(begin);
    vector<vector<bool>> graph = get_graph(words);
    vector<int> visited(words.size(), -1);
    return bfs(graph, visited, words, target);
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def can_convert(word1, word2):
    cnt = 0
    for i in range(len(word1)):
        if word1[i] != word2[i]:
            cnt += 1
    return cnt == 1


def solution(begin, target, words):
    if begin not in words:
        words.append(begin)
        
    N = len(words)
    graph = [[False for _ in range(N)] for _ in range(N)]
    for i in range(N - 1):
        for j in range(i + 1, N):
            if can_convert(words[i], words[j]):
                graph[i][j] = True
                graph[j][i] = True

    q = deque([N - 1])
    visited = [-1 for _ in range(N)]
    visited[N - 1] = 0
    while q:
        v = q.popleft()
        
        if words[v] == target:
            return visited[v]
        
        for w in range(N):
            if graph[v][w] and visited[w] == -1:
                visited[w] = visited[v] + 1
                q.append(w)
        
    return 0